jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/WriteTarFileStep.java

```diff
@@ -36,75 +30,42 @@
 import java.util.concurrent.Callable;
 import java.util.concurrent.ExecutionException;
 
-public class WriteTarFileStep implements AsyncStep<BuildResult>, Callable<BuildResult> {
+public class WriteTarFileStep implements Callable<BuildResult> {
 
-  private final ListeningExecutorService listeningExecutorService;
   private final BuildConfiguration buildConfiguration;
   private final ProgressEventDispatcher.Factory progressEventDispatcherFactory;
 
   private final Path outputPath;
-  private final PullAndCacheBaseImageLayersStep pullAndCacheBaseImageLayersStep;
-  private final ImmutableList<BuildAndCacheApplicationLayerStep> buildAndCacheApplicationLayerSteps;
-  private final BuildImageStep buildImageStep;
-
-  private final ListenableFuture<BuildResult> listenableFuture;
+  private final Image builtImage;
 
   WriteTarFileStep(
-      ListeningExecutorService listeningExecutorService,
       BuildConfiguration buildConfiguration,
       ProgressEventDispatcher.Factory progressEventDispatcherFactory,
       Path outputPath,
-      PullAndCacheBaseImageLayersStep pullAndCacheBaseImageLayersStep,
-      ImmutableList<BuildAndCacheApplicationLayerStep> buildAndCacheApplicationLayerSteps,
-      BuildImageStep buildImageStep) {
-    this.listeningExecutorService = listeningExecutorService;
+      Image builtImage) {
     this.buildConfiguration = buildConfiguration;
     this.progressEventDispatcherFactory = progressEventDispatcherFactory;
     this.outputPath = outputPath;
-    this.pullAndCacheBaseImageLayersStep = pullAndCacheBaseImageLayersStep;
-    this.buildAndCacheApplicationLayerSteps = buildAndCacheApplicationLayerSteps;
-    this.buildImageStep = buildImageStep;
-
-    listenableFuture =
-        AsyncDependencies.using(listeningExecutorService)
-            .addStep(pullAndCacheBaseImageLayersStep)
-            .addStep(buildImageStep)
-            .whenAllSucceed(this);
-  }
-
-  @Override
-  public ListenableFuture<BuildResult> getFuture() {
-    return listenableFuture;
+    this.builtImage = builtImage;
   }
 
   @Override
-  public BuildResult call() throws ExecutionException, InterruptedException {
-    return AsyncDependencies.using(listeningExecutorService)
-        .addSteps(NonBlockingSteps.get(pullAndCacheBaseImageLayersStep))
-        .addSteps(buildAndCacheApplicationLayerSteps)
-        .addStep(NonBlockingSteps.get(buildImageStep))
-        .whenAllSucceed(this::writeTarFile)
-        .get();
```

Comment:
Confirming: the core logic can start only after
1. the base image layers are downloaded and cached;
1. the application layers are built and cached; and
1. the target `Image` instance is created and available.

Note building a target `Image` already requires 1 and 2.